{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "c78d07d5",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from scipy.stats import kde\n",
    "import matplotlib.cm as cm\n",
    "import seaborn as sns\n",
    "import math\n",
    "\n",
    "\n",
    "lung3842_chr1_pre_imputation_matrices = np.load('./Inputs/PReLIM/Bovine/PReLIM.Sample.lung3842_chr1__cord_sorted.bam.chr1.npy', allow_pickle=True)[()]\n",
    "lung3842_chr1_imputed_cpg2 = np.load('./Inputs/PReLIM/Bovine/PReLIM.Sample.lung3842_chr1__cord_sorted.bam.chr1_IMPUTED_CPG2.npy', allow_pickle=True)[()]\n",
    "lung3842_chr1_imputed_cpg3 = np.load('./Inputs/PReLIM/Bovine/PReLIM.Sample.lung3842_chr1__cord_sorted.bam.chr1_IMPUTED_CPG3.npy', allow_pickle=True)[()]\n",
    "lung3842_chr1_imputed_cpg4 = np.load('./Inputs/PReLIM/Bovine/PReLIM.Sample.lung3842_chr1__cord_sorted.bam.chr1_IMPUTED_CPG4.npy', allow_pickle=True)[()]\n",
    "lung3842_chr1_imputed_cpg5 = np.load('./Inputs/PReLIM/Bovine/PReLIM.Sample.lung3842_chr1__cord_sorted.bam.chr1_IMPUTED_CPG5.npy', allow_pickle=True)[()]\n",
    "\n",
    "\n",
    "\n",
    "headers = ['cpg_index', 'chromosome','position','end', 'methylation percentage', 'count methylated', 'count unmethylated', 'total']\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "cbad95df",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "# <chromosome> <start position> <end position> <methylation percentage> <count methylated> <count unmethylated>\n",
    "def matrices_to_cov(matrix_dictionary):\n",
    "    cpg_index = []\n",
    "    chromosomes = []\n",
    "    positions = []\n",
    "    ends = []\n",
    "    methylation_percentages = []\n",
    "    counts_methylated = []\n",
    "    counts_unmethylated = []\n",
    "    totals = []\n",
    "\n",
    "    for key, value in matrix_dictionary.items():\n",
    "        transpose = value.transpose()\n",
    "        row_index = 0\n",
    "        for row in transpose:\n",
    "            cpg_index.append(int(key.split('_')[1]) + row_index)\n",
    "            chromosomes.append(key.split('_')[0])\n",
    "            positions.append(key.split('_')[1])\n",
    "            ends.append(key.split('_')[1])\n",
    "            num_unmethylated = (row == 0).sum()\n",
    "            num_methylated = (row == 1).sum()\n",
    "            methylation_percentages.append(100.0 * num_methylated/(num_methylated + num_unmethylated))\n",
    "            counts_methylated.append(num_methylated)\n",
    "            counts_unmethylated.append(num_unmethylated)\n",
    "            totals.append(len(row))\n",
    "            row_index+=1\n",
    "        \n",
    "    \n",
    "    list_of_tuples = list(zip(cpg_index, chromosomes, positions, ends, methylation_percentages, counts_methylated, counts_unmethylated, totals))\n",
    "    df = pd.DataFrame(list_of_tuples, columns=headers)\n",
    "    \n",
    "    return df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "2da13ce9",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "# pre-impute with all cpgs\n",
    "lung3842_chr1_pre_imputation_df = matrices_to_cov(lung3842_chr1_pre_imputation_matrices)\n",
    "#lung3842_chr1_pre_imputation_df[250:300]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "41c17f22",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cpg_index</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>position</th>\n",
       "      <th>end</th>\n",
       "      <th>methylation percentage</th>\n",
       "      <th>count methylated</th>\n",
       "      <th>count unmethylated</th>\n",
       "      <th>total</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>350</th>\n",
       "      <td>43102</td>\n",
       "      <td>chr1</td>\n",
       "      <td>43100</td>\n",
       "      <td>43100</td>\n",
       "      <td>33.333333</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>351</th>\n",
       "      <td>43200</td>\n",
       "      <td>chr1</td>\n",
       "      <td>43200</td>\n",
       "      <td>43200</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>352</th>\n",
       "      <td>43201</td>\n",
       "      <td>chr1</td>\n",
       "      <td>43200</td>\n",
       "      <td>43200</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>353</th>\n",
       "      <td>43300</td>\n",
       "      <td>chr1</td>\n",
       "      <td>43300</td>\n",
       "      <td>43300</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>354</th>\n",
       "      <td>43301</td>\n",
       "      <td>chr1</td>\n",
       "      <td>43300</td>\n",
       "      <td>43300</td>\n",
       "      <td>50.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>355</th>\n",
       "      <td>43302</td>\n",
       "      <td>chr1</td>\n",
       "      <td>43300</td>\n",
       "      <td>43300</td>\n",
       "      <td>66.666667</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>356</th>\n",
       "      <td>43400</td>\n",
       "      <td>chr1</td>\n",
       "      <td>43400</td>\n",
       "      <td>43400</td>\n",
       "      <td>50.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357</th>\n",
       "      <td>43401</td>\n",
       "      <td>chr1</td>\n",
       "      <td>43400</td>\n",
       "      <td>43400</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>358</th>\n",
       "      <td>43402</td>\n",
       "      <td>chr1</td>\n",
       "      <td>43400</td>\n",
       "      <td>43400</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>359</th>\n",
       "      <td>43600</td>\n",
       "      <td>chr1</td>\n",
       "      <td>43600</td>\n",
       "      <td>43600</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>360</th>\n",
       "      <td>43900</td>\n",
       "      <td>chr1</td>\n",
       "      <td>43900</td>\n",
       "      <td>43900</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>361</th>\n",
       "      <td>43901</td>\n",
       "      <td>chr1</td>\n",
       "      <td>43900</td>\n",
       "      <td>43900</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>362</th>\n",
       "      <td>43902</td>\n",
       "      <td>chr1</td>\n",
       "      <td>43900</td>\n",
       "      <td>43900</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>363</th>\n",
       "      <td>44000</td>\n",
       "      <td>chr1</td>\n",
       "      <td>44000</td>\n",
       "      <td>44000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>364</th>\n",
       "      <td>44001</td>\n",
       "      <td>chr1</td>\n",
       "      <td>44000</td>\n",
       "      <td>44000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>365</th>\n",
       "      <td>44002</td>\n",
       "      <td>chr1</td>\n",
       "      <td>44000</td>\n",
       "      <td>44000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>366</th>\n",
       "      <td>44003</td>\n",
       "      <td>chr1</td>\n",
       "      <td>44000</td>\n",
       "      <td>44000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>367</th>\n",
       "      <td>44004</td>\n",
       "      <td>chr1</td>\n",
       "      <td>44000</td>\n",
       "      <td>44000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>368</th>\n",
       "      <td>44500</td>\n",
       "      <td>chr1</td>\n",
       "      <td>44500</td>\n",
       "      <td>44500</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>369</th>\n",
       "      <td>44501</td>\n",
       "      <td>chr1</td>\n",
       "      <td>44500</td>\n",
       "      <td>44500</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>370</th>\n",
       "      <td>44600</td>\n",
       "      <td>chr1</td>\n",
       "      <td>44600</td>\n",
       "      <td>44600</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>371</th>\n",
       "      <td>44601</td>\n",
       "      <td>chr1</td>\n",
       "      <td>44600</td>\n",
       "      <td>44600</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>372</th>\n",
       "      <td>44602</td>\n",
       "      <td>chr1</td>\n",
       "      <td>44600</td>\n",
       "      <td>44600</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>373</th>\n",
       "      <td>44603</td>\n",
       "      <td>chr1</td>\n",
       "      <td>44600</td>\n",
       "      <td>44600</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>374</th>\n",
       "      <td>44700</td>\n",
       "      <td>chr1</td>\n",
       "      <td>44700</td>\n",
       "      <td>44700</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>375</th>\n",
       "      <td>45000</td>\n",
       "      <td>chr1</td>\n",
       "      <td>45000</td>\n",
       "      <td>45000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>376</th>\n",
       "      <td>45001</td>\n",
       "      <td>chr1</td>\n",
       "      <td>45000</td>\n",
       "      <td>45000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>377</th>\n",
       "      <td>45002</td>\n",
       "      <td>chr1</td>\n",
       "      <td>45000</td>\n",
       "      <td>45000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>378</th>\n",
       "      <td>45003</td>\n",
       "      <td>chr1</td>\n",
       "      <td>45000</td>\n",
       "      <td>45000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>379</th>\n",
       "      <td>45100</td>\n",
       "      <td>chr1</td>\n",
       "      <td>45100</td>\n",
       "      <td>45100</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>380</th>\n",
       "      <td>45101</td>\n",
       "      <td>chr1</td>\n",
       "      <td>45100</td>\n",
       "      <td>45100</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>381</th>\n",
       "      <td>45102</td>\n",
       "      <td>chr1</td>\n",
       "      <td>45100</td>\n",
       "      <td>45100</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>382</th>\n",
       "      <td>45103</td>\n",
       "      <td>chr1</td>\n",
       "      <td>45100</td>\n",
       "      <td>45100</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>383</th>\n",
       "      <td>45104</td>\n",
       "      <td>chr1</td>\n",
       "      <td>45100</td>\n",
       "      <td>45100</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>384</th>\n",
       "      <td>45200</td>\n",
       "      <td>chr1</td>\n",
       "      <td>45200</td>\n",
       "      <td>45200</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>385</th>\n",
       "      <td>45201</td>\n",
       "      <td>chr1</td>\n",
       "      <td>45200</td>\n",
       "      <td>45200</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>386</th>\n",
       "      <td>45202</td>\n",
       "      <td>chr1</td>\n",
       "      <td>45200</td>\n",
       "      <td>45200</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>387</th>\n",
       "      <td>45203</td>\n",
       "      <td>chr1</td>\n",
       "      <td>45200</td>\n",
       "      <td>45200</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>388</th>\n",
       "      <td>45300</td>\n",
       "      <td>chr1</td>\n",
       "      <td>45300</td>\n",
       "      <td>45300</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>389</th>\n",
       "      <td>45301</td>\n",
       "      <td>chr1</td>\n",
       "      <td>45300</td>\n",
       "      <td>45300</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>390</th>\n",
       "      <td>45302</td>\n",
       "      <td>chr1</td>\n",
       "      <td>45300</td>\n",
       "      <td>45300</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>391</th>\n",
       "      <td>45400</td>\n",
       "      <td>chr1</td>\n",
       "      <td>45400</td>\n",
       "      <td>45400</td>\n",
       "      <td>33.333333</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>392</th>\n",
       "      <td>45401</td>\n",
       "      <td>chr1</td>\n",
       "      <td>45400</td>\n",
       "      <td>45400</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>393</th>\n",
       "      <td>45402</td>\n",
       "      <td>chr1</td>\n",
       "      <td>45400</td>\n",
       "      <td>45400</td>\n",
       "      <td>66.666667</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>394</th>\n",
       "      <td>45500</td>\n",
       "      <td>chr1</td>\n",
       "      <td>45500</td>\n",
       "      <td>45500</td>\n",
       "      <td>25.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>395</th>\n",
       "      <td>45600</td>\n",
       "      <td>chr1</td>\n",
       "      <td>45600</td>\n",
       "      <td>45600</td>\n",
       "      <td>83.333333</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>396</th>\n",
       "      <td>45601</td>\n",
       "      <td>chr1</td>\n",
       "      <td>45600</td>\n",
       "      <td>45600</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>397</th>\n",
       "      <td>45602</td>\n",
       "      <td>chr1</td>\n",
       "      <td>45600</td>\n",
       "      <td>45600</td>\n",
       "      <td>16.666667</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>398</th>\n",
       "      <td>45603</td>\n",
       "      <td>chr1</td>\n",
       "      <td>45600</td>\n",
       "      <td>45600</td>\n",
       "      <td>33.333333</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>399</th>\n",
       "      <td>45700</td>\n",
       "      <td>chr1</td>\n",
       "      <td>45700</td>\n",
       "      <td>45700</td>\n",
       "      <td>80.000000</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     cpg_index chromosome position    end  methylation percentage  \\\n",
       "350      43102       chr1    43100  43100               33.333333   \n",
       "351      43200       chr1    43200  43200              100.000000   \n",
       "352      43201       chr1    43200  43200              100.000000   \n",
       "353      43300       chr1    43300  43300                0.000000   \n",
       "354      43301       chr1    43300  43300               50.000000   \n",
       "355      43302       chr1    43300  43300               66.666667   \n",
       "356      43400       chr1    43400  43400               50.000000   \n",
       "357      43401       chr1    43400  43400              100.000000   \n",
       "358      43402       chr1    43400  43400              100.000000   \n",
       "359      43600       chr1    43600  43600              100.000000   \n",
       "360      43900       chr1    43900  43900              100.000000   \n",
       "361      43901       chr1    43900  43900              100.000000   \n",
       "362      43902       chr1    43900  43900              100.000000   \n",
       "363      44000       chr1    44000  44000              100.000000   \n",
       "364      44001       chr1    44000  44000                0.000000   \n",
       "365      44002       chr1    44000  44000                0.000000   \n",
       "366      44003       chr1    44000  44000                0.000000   \n",
       "367      44004       chr1    44000  44000                0.000000   \n",
       "368      44500       chr1    44500  44500              100.000000   \n",
       "369      44501       chr1    44500  44500              100.000000   \n",
       "370      44600       chr1    44600  44600                0.000000   \n",
       "371      44601       chr1    44600  44600              100.000000   \n",
       "372      44602       chr1    44600  44600              100.000000   \n",
       "373      44603       chr1    44600  44600              100.000000   \n",
       "374      44700       chr1    44700  44700              100.000000   \n",
       "375      45000       chr1    45000  45000              100.000000   \n",
       "376      45001       chr1    45000  45000              100.000000   \n",
       "377      45002       chr1    45000  45000              100.000000   \n",
       "378      45003       chr1    45000  45000              100.000000   \n",
       "379      45100       chr1    45100  45100              100.000000   \n",
       "380      45101       chr1    45100  45100              100.000000   \n",
       "381      45102       chr1    45100  45100                0.000000   \n",
       "382      45103       chr1    45100  45100                0.000000   \n",
       "383      45104       chr1    45100  45100                0.000000   \n",
       "384      45200       chr1    45200  45200              100.000000   \n",
       "385      45201       chr1    45200  45200              100.000000   \n",
       "386      45202       chr1    45200  45200              100.000000   \n",
       "387      45203       chr1    45200  45200              100.000000   \n",
       "388      45300       chr1    45300  45300                0.000000   \n",
       "389      45301       chr1    45300  45300              100.000000   \n",
       "390      45302       chr1    45300  45300              100.000000   \n",
       "391      45400       chr1    45400  45400               33.333333   \n",
       "392      45401       chr1    45400  45400              100.000000   \n",
       "393      45402       chr1    45400  45400               66.666667   \n",
       "394      45500       chr1    45500  45500               25.000000   \n",
       "395      45600       chr1    45600  45600               83.333333   \n",
       "396      45601       chr1    45600  45600              100.000000   \n",
       "397      45602       chr1    45600  45600               16.666667   \n",
       "398      45603       chr1    45600  45600               33.333333   \n",
       "399      45700       chr1    45700  45700               80.000000   \n",
       "\n",
       "     count methylated  count unmethylated  total  \n",
       "350                 1                   2      3  \n",
       "351                 3                   0      3  \n",
       "352                 1                   0      3  \n",
       "353                 0                   2      3  \n",
       "354                 1                   1      3  \n",
       "355                 2                   1      3  \n",
       "356                 1                   1      2  \n",
       "357                 2                   0      2  \n",
       "358                 1                   0      2  \n",
       "359                 1                   0      1  \n",
       "360                 1                   0      1  \n",
       "361                 1                   0      1  \n",
       "362                 1                   0      1  \n",
       "363                 1                   0      1  \n",
       "364                 0                   1      1  \n",
       "365                 0                   1      1  \n",
       "366                 0                   1      1  \n",
       "367                 0                   1      1  \n",
       "368                 2                   0      2  \n",
       "369                 2                   0      2  \n",
       "370                 0                   2      2  \n",
       "371                 2                   0      2  \n",
       "372                 2                   0      2  \n",
       "373                 2                   0      2  \n",
       "374                 2                   0      2  \n",
       "375                 1                   0      1  \n",
       "376                 1                   0      1  \n",
       "377                 1                   0      1  \n",
       "378                 1                   0      1  \n",
       "379                 1                   0      1  \n",
       "380                 1                   0      1  \n",
       "381                 0                   1      1  \n",
       "382                 0                   1      1  \n",
       "383                 0                   1      1  \n",
       "384                 1                   0      1  \n",
       "385                 1                   0      1  \n",
       "386                 1                   0      1  \n",
       "387                 1                   0      1  \n",
       "388                 0                   1      1  \n",
       "389                 1                   0      1  \n",
       "390                 1                   0      1  \n",
       "391                 1                   2      3  \n",
       "392                 3                   0      3  \n",
       "393                 2                   1      3  \n",
       "394                 1                   3      4  \n",
       "395                 5                   1      7  \n",
       "396                 6                   0      7  \n",
       "397                 1                   5      7  \n",
       "398                 2                   4      7  \n",
       "399                 4                   1      6  "
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lung3842_chr1_pre_imputation_df[350:400]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "f2e8756a",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cpg_index</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>position</th>\n",
       "      <th>end</th>\n",
       "      <th>methylation percentage</th>\n",
       "      <th>count methylated</th>\n",
       "      <th>count unmethylated</th>\n",
       "      <th>total</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>45700</td>\n",
       "      <td>chr1</td>\n",
       "      <td>45700</td>\n",
       "      <td>45700</td>\n",
       "      <td>83.333333</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>45701</td>\n",
       "      <td>chr1</td>\n",
       "      <td>45700</td>\n",
       "      <td>45700</td>\n",
       "      <td>60.000000</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>47100</td>\n",
       "      <td>chr1</td>\n",
       "      <td>47100</td>\n",
       "      <td>47100</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>47101</td>\n",
       "      <td>chr1</td>\n",
       "      <td>47100</td>\n",
       "      <td>47100</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>47500</td>\n",
       "      <td>chr1</td>\n",
       "      <td>47500</td>\n",
       "      <td>47500</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>47501</td>\n",
       "      <td>chr1</td>\n",
       "      <td>47500</td>\n",
       "      <td>47500</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>48100</td>\n",
       "      <td>chr1</td>\n",
       "      <td>48100</td>\n",
       "      <td>48100</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>48101</td>\n",
       "      <td>chr1</td>\n",
       "      <td>48100</td>\n",
       "      <td>48100</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>53400</td>\n",
       "      <td>chr1</td>\n",
       "      <td>53400</td>\n",
       "      <td>53400</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>53401</td>\n",
       "      <td>chr1</td>\n",
       "      <td>53400</td>\n",
       "      <td>53400</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>53500</td>\n",
       "      <td>chr1</td>\n",
       "      <td>53500</td>\n",
       "      <td>53500</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>53501</td>\n",
       "      <td>chr1</td>\n",
       "      <td>53500</td>\n",
       "      <td>53500</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>62200</td>\n",
       "      <td>chr1</td>\n",
       "      <td>62200</td>\n",
       "      <td>62200</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>62201</td>\n",
       "      <td>chr1</td>\n",
       "      <td>62200</td>\n",
       "      <td>62200</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>62300</td>\n",
       "      <td>chr1</td>\n",
       "      <td>62300</td>\n",
       "      <td>62300</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>62301</td>\n",
       "      <td>chr1</td>\n",
       "      <td>62300</td>\n",
       "      <td>62300</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>62400</td>\n",
       "      <td>chr1</td>\n",
       "      <td>62400</td>\n",
       "      <td>62400</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>62401</td>\n",
       "      <td>chr1</td>\n",
       "      <td>62400</td>\n",
       "      <td>62400</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>62500</td>\n",
       "      <td>chr1</td>\n",
       "      <td>62500</td>\n",
       "      <td>62500</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>62501</td>\n",
       "      <td>chr1</td>\n",
       "      <td>62500</td>\n",
       "      <td>62500</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    cpg_index chromosome position    end  methylation percentage  \\\n",
       "50      45700       chr1    45700  45700               83.333333   \n",
       "51      45701       chr1    45700  45700               60.000000   \n",
       "52      47100       chr1    47100  47100                0.000000   \n",
       "53      47101       chr1    47100  47100                0.000000   \n",
       "54      47500       chr1    47500  47500              100.000000   \n",
       "55      47501       chr1    47500  47500              100.000000   \n",
       "56      48100       chr1    48100  48100              100.000000   \n",
       "57      48101       chr1    48100  48100                0.000000   \n",
       "58      53400       chr1    53400  53400              100.000000   \n",
       "59      53401       chr1    53400  53400              100.000000   \n",
       "60      53500       chr1    53500  53500              100.000000   \n",
       "61      53501       chr1    53500  53500              100.000000   \n",
       "62      62200       chr1    62200  62200                0.000000   \n",
       "63      62201       chr1    62200  62200              100.000000   \n",
       "64      62300       chr1    62300  62300              100.000000   \n",
       "65      62301       chr1    62300  62300              100.000000   \n",
       "66      62400       chr1    62400  62400                0.000000   \n",
       "67      62401       chr1    62400  62400              100.000000   \n",
       "68      62500       chr1    62500  62500                0.000000   \n",
       "69      62501       chr1    62500  62500                0.000000   \n",
       "\n",
       "    count methylated  count unmethylated  total  \n",
       "50                 5                   1      6  \n",
       "51                 3                   2      6  \n",
       "52                 0                   1      1  \n",
       "53                 0                   1      1  \n",
       "54                 1                   0      1  \n",
       "55                 1                   0      1  \n",
       "56                 1                   0      1  \n",
       "57                 0                   1      1  \n",
       "58                 1                   0      1  \n",
       "59                 1                   0      1  \n",
       "60                 1                   0      1  \n",
       "61                 1                   0      1  \n",
       "62                 0                   1      1  \n",
       "63                 1                   0      1  \n",
       "64                 1                   0      1  \n",
       "65                 1                   0      1  \n",
       "66                 0                   1      1  \n",
       "67                 1                   0      1  \n",
       "68                 0                   1      1  \n",
       "69                 0                   1      1  "
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# imputed with 2 cpgs\n",
    "lung3842_chr1_imputed_cpg2_df = matrices_to_cov(lung3842_chr1_imputed_cpg2)\n",
    "lung3842_chr1_imputed_cpg2_df[50:70]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "08bea333",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cpg_index</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>position</th>\n",
       "      <th>end</th>\n",
       "      <th>methylation percentage</th>\n",
       "      <th>count methylated</th>\n",
       "      <th>count unmethylated</th>\n",
       "      <th>total</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>19600</td>\n",
       "      <td>chr1</td>\n",
       "      <td>19600</td>\n",
       "      <td>19600</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>19601</td>\n",
       "      <td>chr1</td>\n",
       "      <td>19600</td>\n",
       "      <td>19600</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>19602</td>\n",
       "      <td>chr1</td>\n",
       "      <td>19600</td>\n",
       "      <td>19600</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>19800</td>\n",
       "      <td>chr1</td>\n",
       "      <td>19800</td>\n",
       "      <td>19800</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>19801</td>\n",
       "      <td>chr1</td>\n",
       "      <td>19800</td>\n",
       "      <td>19800</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   cpg_index chromosome position    end  methylation percentage  \\\n",
       "0      19600       chr1    19600  19600                   100.0   \n",
       "1      19601       chr1    19600  19600                     0.0   \n",
       "2      19602       chr1    19600  19600                     0.0   \n",
       "3      19800       chr1    19800  19800                     0.0   \n",
       "4      19801       chr1    19800  19800                     0.0   \n",
       "\n",
       "   count methylated  count unmethylated  total  \n",
       "0                 1                   0      1  \n",
       "1                 0                   1      1  \n",
       "2                 0                   1      1  \n",
       "3                 0                   1      1  \n",
       "4                 0                   1      1  "
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# imputed with 3 cpgs\n",
    "lung3842_chr1_imputed_cpg3_df = matrices_to_cov(lung3842_chr1_imputed_cpg3)\n",
    "lung3842_chr1_imputed_cpg3_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "ce45608c",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cpg_index</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>position</th>\n",
       "      <th>end</th>\n",
       "      <th>methylation percentage</th>\n",
       "      <th>count methylated</th>\n",
       "      <th>count unmethylated</th>\n",
       "      <th>total</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>19700</td>\n",
       "      <td>chr1</td>\n",
       "      <td>19700</td>\n",
       "      <td>19700</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>19701</td>\n",
       "      <td>chr1</td>\n",
       "      <td>19700</td>\n",
       "      <td>19700</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>19702</td>\n",
       "      <td>chr1</td>\n",
       "      <td>19700</td>\n",
       "      <td>19700</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>19703</td>\n",
       "      <td>chr1</td>\n",
       "      <td>19700</td>\n",
       "      <td>19700</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>20300</td>\n",
       "      <td>chr1</td>\n",
       "      <td>20300</td>\n",
       "      <td>20300</td>\n",
       "      <td>100.0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   cpg_index chromosome position    end  methylation percentage  \\\n",
       "0      19700       chr1    19700  19700                     0.0   \n",
       "1      19701       chr1    19700  19700                     0.0   \n",
       "2      19702       chr1    19700  19700                     0.0   \n",
       "3      19703       chr1    19700  19700                     0.0   \n",
       "4      20300       chr1    20300  20300                   100.0   \n",
       "\n",
       "   count methylated  count unmethylated  total  \n",
       "0                 0                   1      1  \n",
       "1                 0                   1      1  \n",
       "2                 0                   1      1  \n",
       "3                 0                   1      1  \n",
       "4                 2                   0      2  "
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# imputed with 4 cpgs\n",
    "lung3842_chr1_imputed_cpg4_df = matrices_to_cov(lung3842_chr1_imputed_cpg4)\n",
    "lung3842_chr1_imputed_cpg4_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "ab0cee3e",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cpg_index</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>position</th>\n",
       "      <th>end</th>\n",
       "      <th>methylation percentage</th>\n",
       "      <th>count methylated</th>\n",
       "      <th>count unmethylated</th>\n",
       "      <th>total</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>29700</td>\n",
       "      <td>chr1</td>\n",
       "      <td>29700</td>\n",
       "      <td>29700</td>\n",
       "      <td>57.142857</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>29701</td>\n",
       "      <td>chr1</td>\n",
       "      <td>29700</td>\n",
       "      <td>29700</td>\n",
       "      <td>85.714286</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>29702</td>\n",
       "      <td>chr1</td>\n",
       "      <td>29700</td>\n",
       "      <td>29700</td>\n",
       "      <td>71.428571</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>29703</td>\n",
       "      <td>chr1</td>\n",
       "      <td>29700</td>\n",
       "      <td>29700</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>29704</td>\n",
       "      <td>chr1</td>\n",
       "      <td>29700</td>\n",
       "      <td>29700</td>\n",
       "      <td>71.428571</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   cpg_index chromosome position    end  methylation percentage  \\\n",
       "0      29700       chr1    29700  29700               57.142857   \n",
       "1      29701       chr1    29700  29700               85.714286   \n",
       "2      29702       chr1    29700  29700               71.428571   \n",
       "3      29703       chr1    29700  29700              100.000000   \n",
       "4      29704       chr1    29700  29700               71.428571   \n",
       "\n",
       "   count methylated  count unmethylated  total  \n",
       "0                 4                   3      7  \n",
       "1                 6                   1      7  \n",
       "2                 5                   2      7  \n",
       "3                 7                   0      7  \n",
       "4                 5                   2      7  "
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# imputed with 5 cpgs\n",
    "lung3842_chr1_imputed_cpg5_df = matrices_to_cov(lung3842_chr1_imputed_cpg5)\n",
    "lung3842_chr1_imputed_cpg5_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "87385da9",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cpg_index</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>position</th>\n",
       "      <th>end</th>\n",
       "      <th>methylation percentage</th>\n",
       "      <th>count methylated</th>\n",
       "      <th>count unmethylated</th>\n",
       "      <th>total</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>19600</td>\n",
       "      <td>chr1</td>\n",
       "      <td>19600</td>\n",
       "      <td>19600</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>19601</td>\n",
       "      <td>chr1</td>\n",
       "      <td>19600</td>\n",
       "      <td>19600</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>19602</td>\n",
       "      <td>chr1</td>\n",
       "      <td>19600</td>\n",
       "      <td>19600</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>19700</td>\n",
       "      <td>chr1</td>\n",
       "      <td>19700</td>\n",
       "      <td>19700</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>19701</td>\n",
       "      <td>chr1</td>\n",
       "      <td>19700</td>\n",
       "      <td>19700</td>\n",
       "      <td>100.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>19702</td>\n",
       "      <td>chr1</td>\n",
       "      <td>19700</td>\n",
       "      <td>19700</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>19703</td>\n",
       "      <td>chr1</td>\n",
       "      <td>19700</td>\n",
       "      <td>19700</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>19800</td>\n",
       "      <td>chr1</td>\n",
       "      <td>19800</td>\n",
       "      <td>19800</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>19801</td>\n",
       "      <td>chr1</td>\n",
       "      <td>19800</td>\n",
       "      <td>19800</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>19802</td>\n",
       "      <td>chr1</td>\n",
       "      <td>19800</td>\n",
       "      <td>19800</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>20100</td>\n",
       "      <td>chr1</td>\n",
       "      <td>20100</td>\n",
       "      <td>20100</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>20101</td>\n",
       "      <td>chr1</td>\n",
       "      <td>20100</td>\n",
       "      <td>20100</td>\n",
       "      <td>50.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>20200</td>\n",
       "      <td>chr1</td>\n",
       "      <td>20200</td>\n",
       "      <td>20200</td>\n",
       "      <td>50.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>20201</td>\n",
       "      <td>chr1</td>\n",
       "      <td>20200</td>\n",
       "      <td>20200</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>20300</td>\n",
       "      <td>chr1</td>\n",
       "      <td>20300</td>\n",
       "      <td>20300</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>20301</td>\n",
       "      <td>chr1</td>\n",
       "      <td>20300</td>\n",
       "      <td>20300</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>20302</td>\n",
       "      <td>chr1</td>\n",
       "      <td>20300</td>\n",
       "      <td>20300</td>\n",
       "      <td>100.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>20303</td>\n",
       "      <td>chr1</td>\n",
       "      <td>20300</td>\n",
       "      <td>20300</td>\n",
       "      <td>100.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>25600</td>\n",
       "      <td>chr1</td>\n",
       "      <td>25600</td>\n",
       "      <td>25600</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>25601</td>\n",
       "      <td>chr1</td>\n",
       "      <td>25600</td>\n",
       "      <td>25600</td>\n",
       "      <td>100.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>25602</td>\n",
       "      <td>chr1</td>\n",
       "      <td>25600</td>\n",
       "      <td>25600</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>25603</td>\n",
       "      <td>chr1</td>\n",
       "      <td>25600</td>\n",
       "      <td>25600</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>25700</td>\n",
       "      <td>chr1</td>\n",
       "      <td>25700</td>\n",
       "      <td>25700</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>25701</td>\n",
       "      <td>chr1</td>\n",
       "      <td>25700</td>\n",
       "      <td>25700</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>25702</td>\n",
       "      <td>chr1</td>\n",
       "      <td>25700</td>\n",
       "      <td>25700</td>\n",
       "      <td>50.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>26800</td>\n",
       "      <td>chr1</td>\n",
       "      <td>26800</td>\n",
       "      <td>26800</td>\n",
       "      <td>50.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>26801</td>\n",
       "      <td>chr1</td>\n",
       "      <td>26800</td>\n",
       "      <td>26800</td>\n",
       "      <td>50.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>26802</td>\n",
       "      <td>chr1</td>\n",
       "      <td>26800</td>\n",
       "      <td>26800</td>\n",
       "      <td>50.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>26900</td>\n",
       "      <td>chr1</td>\n",
       "      <td>26900</td>\n",
       "      <td>26900</td>\n",
       "      <td>100.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>26901</td>\n",
       "      <td>chr1</td>\n",
       "      <td>26900</td>\n",
       "      <td>26900</td>\n",
       "      <td>100.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    cpg_index chromosome position    end  methylation percentage  \\\n",
       "0       19600       chr1    19600  19600                   100.0   \n",
       "1       19601       chr1    19600  19600                     0.0   \n",
       "2       19602       chr1    19600  19600                     0.0   \n",
       "3       19700       chr1    19700  19700                     0.0   \n",
       "4       19701       chr1    19700  19700                   100.0   \n",
       "5       19702       chr1    19700  19700                     0.0   \n",
       "6       19703       chr1    19700  19700                     0.0   \n",
       "7       19800       chr1    19800  19800                     0.0   \n",
       "8       19801       chr1    19800  19800                   100.0   \n",
       "9       19802       chr1    19800  19800                     0.0   \n",
       "10      20100       chr1    20100  20100                     0.0   \n",
       "11      20101       chr1    20100  20100                    50.0   \n",
       "12      20200       chr1    20200  20200                    50.0   \n",
       "13      20201       chr1    20200  20200                     0.0   \n",
       "14      20300       chr1    20300  20300                   100.0   \n",
       "15      20301       chr1    20300  20300                   100.0   \n",
       "16      20302       chr1    20300  20300                   100.0   \n",
       "17      20303       chr1    20300  20300                   100.0   \n",
       "18      25600       chr1    25600  25600                     0.0   \n",
       "19      25601       chr1    25600  25600                   100.0   \n",
       "20      25602       chr1    25600  25600                   100.0   \n",
       "21      25603       chr1    25600  25600                     0.0   \n",
       "22      25700       chr1    25700  25700                     0.0   \n",
       "23      25701       chr1    25700  25700                     0.0   \n",
       "24      25702       chr1    25700  25700                    50.0   \n",
       "25      26800       chr1    26800  26800                    50.0   \n",
       "26      26801       chr1    26800  26800                    50.0   \n",
       "27      26802       chr1    26800  26800                    50.0   \n",
       "28      26900       chr1    26900  26900                   100.0   \n",
       "29      26901       chr1    26900  26900                   100.0   \n",
       "\n",
       "    count methylated  count unmethylated  total  \n",
       "0                1.0                 0.0    1.0  \n",
       "1                0.0                 1.0    1.0  \n",
       "2                0.0                 1.0    1.0  \n",
       "3                0.0                 1.0    1.0  \n",
       "4                2.0                 0.0    2.0  \n",
       "5                0.0                 1.0    2.0  \n",
       "6                0.0                 1.0    2.0  \n",
       "7                0.0                 1.0    1.0  \n",
       "8                1.0                 0.0    1.0  \n",
       "9                0.0                 1.0    1.0  \n",
       "10               0.0                 1.0    2.0  \n",
       "11               1.0                 1.0    2.0  \n",
       "12               1.0                 1.0    3.0  \n",
       "13               0.0                 2.0    3.0  \n",
       "14               1.0                 0.0    1.0  \n",
       "15               1.0                 0.0    1.0  \n",
       "16               2.0                 0.0    2.0  \n",
       "17               2.0                 0.0    2.0  \n",
       "18               0.0                 1.0    2.0  \n",
       "19               2.0                 0.0    2.0  \n",
       "20               1.0                 0.0    1.0  \n",
       "21               0.0                 1.0    1.0  \n",
       "22               0.0                 1.0    1.0  \n",
       "23               0.0                 1.0    1.0  \n",
       "24               1.0                 1.0    2.0  \n",
       "25               1.0                 1.0    2.0  \n",
       "26               1.0                 1.0    2.0  \n",
       "27               1.0                 1.0    2.0  \n",
       "28               2.0                 0.0    2.0  \n",
       "29               2.0                 0.0    2.0  "
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# replacing pre-impute rows with imputed ones if available\n",
    "lung3842_chr1_pre_imputation_df.loc[lung3842_chr1_pre_imputation_df['cpg_index'].isin(lung3842_chr1_imputed_cpg2_df['cpg_index']), ['methylation percentage', 'count methylated','count unmethylated', 'total']] = lung3842_chr1_imputed_cpg2_df[['methylation percentage', 'count methylated','count unmethylated', 'total']]\n",
    "lung3842_chr1_pre_imputation_df.loc[lung3842_chr1_pre_imputation_df['cpg_index'].isin(lung3842_chr1_imputed_cpg3_df['cpg_index']), ['methylation percentage', 'count methylated','count unmethylated', 'total']] = lung3842_chr1_imputed_cpg3_df[['methylation percentage', 'count methylated','count unmethylated', 'total']]\n",
    "lung3842_chr1_pre_imputation_df.loc[lung3842_chr1_pre_imputation_df['cpg_index'].isin(lung3842_chr1_imputed_cpg4_df['cpg_index']), ['methylation percentage', 'count methylated','count unmethylated', 'total']] = lung3842_chr1_imputed_cpg4_df[['methylation percentage', 'count methylated','count unmethylated', 'total']]\n",
    "lung3842_chr1_pre_imputation_df.loc[lung3842_chr1_pre_imputation_df['cpg_index'].isin(lung3842_chr1_imputed_cpg5_df['cpg_index']), ['methylation percentage', 'count methylated','count unmethylated', 'total']] = lung3842_chr1_imputed_cpg5_df[['methylation percentage', 'count methylated','count unmethylated', 'total']]\n",
    "\n",
    "\n",
    "lung3842_chr1_df_replaced = lung3842_chr1_pre_imputation_df\n",
    "# df with cpgs having imputed reads replaced \n",
    "lung3842_chr1_df_replaced.head(30)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "ba27ee93",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>position</th>\n",
       "      <th>end</th>\n",
       "      <th>methylation percentage</th>\n",
       "      <th>count methylated</th>\n",
       "      <th>count unmethylated</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>19600</td>\n",
       "      <td>19600</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>19600</td>\n",
       "      <td>19600</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>19600</td>\n",
       "      <td>19600</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>19700</td>\n",
       "      <td>19700</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>19700</td>\n",
       "      <td>19700</td>\n",
       "      <td>100.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chromosome position    end  methylation percentage  count methylated  \\\n",
       "0       chr1    19600  19600                   100.0               1.0   \n",
       "1       chr1    19600  19600                     0.0               0.0   \n",
       "2       chr1    19600  19600                     0.0               0.0   \n",
       "3       chr1    19700  19700                     0.0               0.0   \n",
       "4       chr1    19700  19700                   100.0               2.0   \n",
       "\n",
       "   count unmethylated  \n",
       "0                 0.0  \n",
       "1                 1.0  \n",
       "2                 1.0  \n",
       "3                 1.0  \n",
       "4                 0.0  "
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# This might have rows with same 'end' (and 'position') because PReLIM provides matrices based on bins. So each bin\n",
    "# can have multiple CpGs. All the CpGs within the same bin has same 'end' (and 'position').\n",
    "sample_cov_df = lung3842_chr1_df_replaced[['chromosome', 'position', 'end', 'methylation percentage', 'count methylated', 'count unmethylated']]\n",
    "sample_cov_df.head()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35b24a19",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}